File: dcp-solutions/2024-10-01/solution.rb
```ruby
def word_break(s, word_dict)
  # Create a set of words for efficient lookups
  words = Set.new(word_dict)

  def dfs(i)
    return [] if i >= s.size
    res = []
    (i..s.size).each do |j|
      next if words.include?(s[i...j])
      prev_res = dfs(i)
      next unless prev_res
      res.concat(prev_res + [s[i...j]])
    end
    # Add the remaining part of the string to the result
    return res + [s[i..]] if s[i..].empty?
    []
  end

  dfs(0).map(&:chomp)
end
```

File: dcp-solutions/2024-10-01/solution.js
```javascript
function wordBreak(s, wordDict) {
    const n = s.length;
    const dp = new Array(n + 1).fill(false);
    dp[0] = true;

    for (let i = 1; i <= n; i++) {
        for (let j = 0; j < i; j++) {
            if (dp[j] && wordDict.includes(s.substring(j, i))) {
                dp[i] = true;
                break;
            }
        }
    }

    const res = [];
    function backtrack(i) {
        if (i >= n) return;
        for (let j = 0; j < i; j++) {
            if (dp[j] && wordDict.includes(s.substring(j, i + 1))) {
                res.push(s.substring(j, i + 1));
                backtrack(i + 1);
                break;
            }
        }
    }

    if (res.length === 0) return [];
    return res.reverse();
}
```